### DISCUSSION TITLE
Type hints and tooltips for decorators/wrappers/proxy objects

### DISCUSSION THREAD

**Discussion by stereobutter** [microsoft/pyright#4426] [General]
I often use decorators/wrappers/proxy objects to extend classes from third party libraries (instead of inheritance). I find this is usually much easier to implement than mucking with `super()` and taking care to uphold the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle). 

Say I've got the following class in a third party library that I want to extend:

```python
class Obj:
    def foo(self, question: str) -> int:
        """Return the answer to the Ultimate Question of Life, the Universe, and Everything"""
        return 42
```

and I write the following wrapper
```python
class Wrapper:
    def __init__(self, obj: Obj):
        self._obj = obj

    def foo(self, *args, **kwargs):
        print("Hello from Wrapper")
        return self._obj.foo(*args, **kwargs)
```

The problem with this is that one looses both the type hints and the doc string this way. Unfortunately `functool.wraps` also does currently not work to restore these to us so that VS Code is able to display them.
<img width="704" alt="Bildschirmfoto 2023-01-10 um 02 25 27" src="https://user-images.githubusercontent.com/35063695/211441179-6a0f216a-9185-4204-865d-8a3d8ab9cbd2.png">

I saw there are multiple issues about this already, in particular https://github.com/microsoft/pyright/issues/3939, https://github.com/microsoft/pylance-release/issues/442 and https://github.com/python/typeshed/pull/6670. But since I don't want to wait around for typeshed to fix this eventually, I've hacked away and ended up with this

```python
from typing import TypeVar, Sequence, Callable, cast, Any
from typing_extensions import ParamSpec, Concatenate, TypeAlias,
import functools

W = TypeVar("W")
P = ParamSpec("P")
R = TypeVar("R")


def wraps_method(
    wrapped: Callable[Concatenate[Any, P], R],
    assigned: Sequence[str] = ...,
    updated: Sequence[str] = ...,
) -> Callable[[Callable[Concatenate[Any, P], R]], Callable[Concatenate[W, P], R]]:
    return cast(
        Callable[[Callable[Concatenate[Any, P], R]], Callable[Concatenate[W, P], R]],
        functools.wraps,
    )
```

that when used with the wrapper class

```python
class Wrapper:
    def __init__(self, obj: Obj):
        self._obj = obj

    @wraps_method(Obj.foo)
    def foo(self, *args, **kwargs):
        return self._obj.foo(*args, **kwargs)
```

tricks VS Code into displaying the original type hints and doc string ✨ 🚀 

<img width="658" alt="Bildschirmfoto 2023-01-10 um 02 34 51" src="https://user-images.githubusercontent.com/35063695/211441861-dc38deac-eaa2-4dff-9282-d44136820c74.png">

I though this might be of some use for others but I've also got a couple of questions/issues:

* Is this a reliable workaround that I can expect to work in future versions of pyright/pylance? How does pyright/pylance know that it should display the doc string by looking at the type hints?
* Is this use case also addressed and resolved by https://github.com/python/typeshed/pull/6670?
* Can I ignore pyright warning me about `TypeVar "W" appears only once in generic function signaturePylance[reportInvalidTypeVarUse]`?
* When I tried to tidy `wraps_method` up by replacing the type definitions by more compact type aliases everything stopped working. Is this a bug or did I miss something there?
    ```python
    WrappedT: TypeAlias = Callable[Concatenate[Any, P], R]
    WrapperT: TypeAlias = Callable[Concatenate[W, P], R]
    
    
    def wraps_method(
        wrapped: WrappedT,
        assigned: Sequence[str] = ...,
        updated: Sequence[str] = ...,
    ) -> Callable[[WrappedT], WrapperT]:
        return functools.wraps
    ```
    <img width="514" alt="Bildschirmfoto 2023-01-10 um 02 40 57" src="https://user-images.githubusercontent.com/35063695/211442544-e8440946-e60b-408a-9aee-45959c96bfbc.png">

**Comment by erictraut** [comment]
Yes, this is a reliable workaround. Pyright captures the docstring from function/method captured by the ParamSpec and uses it in the resulting (wrapped) callable.

Yes, this use case should be resolved by the `functools.wraps` change in typeshed. That PR has been delayed because of limitations in mypy and other type checkers. I'm hopeful that it will eventually be accepted.

The error about `W` is legitimate. You're using it in a manner where it has no meaning. It might as well be replaced by `Any`. To have meaning, `W` should appear at least once in one of the input parameter type annotations.

The type aliases you're defining are generic type aliases. For example, `WrappedT` has two type parameters. The first is a ParamSpec and the second is a regular TypeVar. When you use the type alias, you must provide corresponding type arguments, but you've failed to do so. The type expression `WrappedT` is equivalent to `WrappedT[..., Any]`, which is probably not what you intended. If you replace `WrappedT` with `WrappedT[P, R]`, it will work as you expect. If you enable the `reportMissingTypeArgument` check, pyright will tell you about these missing type arguments.

**Reply by stereobutter** [reply]
Thanks for your answer 🙇; especially your last paragraph cleared up a gap in my knowledge of how type aliases work that I previously didn't even know about. For anyone that might find their way here what I ended up using is

```python
P = ParamSpec("P")
R = TypeVar("R")
MethodT: TypeAlias = Callable[Concatenate[Any, P], R]

def wraps_method(
    wrapped: MethodT[P, R],
    assigned: Sequence[str] = ...,
    updated: Sequence[str] = ...,
) -> Callable[[MethodT[P, R]], MethodT[P, R]]:
    return cast(Callable[[MethodT[P, R]], MethodT[P, R]], functools.wraps)
```

**Reply by erictraut** [reply]
Yes, generic type aliases are commonly misunderstood in Python. That's partly what is motiving the new syntax proposed in [PEP 695](https://peps.python.org/pep-0695/#generic-type-alias).

**Reply by stereobutter** [reply]
@erictraut I though I was happy with my workaround but some of the methods I want to wrap use `overload`  which apparently doesn't work with `ParamSpec` (see https://github.com/python/mypy/issues/13540). Any ideas?

**Reply by stereobutter** [reply]
This seems to do the trick

```python
def wraps(
    wrapped: T,
    assigned: Sequence[str] = ...,
    updated: Sequence[str] = ...,
) -> Callable[[T], T]:
    return cast(Callable[[Any], T], functools.wraps)
```

although it requires the use of `# type: ignore`  for the definition of the wrapped method

```python
class Obj:
    @overload
    def foo(self, question: str) -> int:
        ...

    @overload
    def foo(self, question: int) -> str:
        ...

    def foo(self, question):
        if isinstance(question, str):
            return int(question)
        else:
            return str(question)


class Wrapper:
    def __init__(self, obj: Obj):
        self._obj = obj

    @wraps(Obj.foo)  # type: ignore
    def foo(self, *args, **kwargs):
        print("hello from Wrapper")
        return self._obj.foo(*args, **kwargs)
    
```
<img width="329" alt="Bildschirmfoto 2023-01-12 um 11 13 41" src="https://user-images.githubusercontent.com/35063695/212039893-f0f07081-3e4b-4c12-b03e-c1ba52cf714a.png">

One has to live with the fact though that the lie is exposed by looking at

<img width="424" alt="Bildschirmfoto 2023-01-12 um 11 13 23" src="https://user-images.githubusercontent.com/35063695/212040001-5cf8fdab-cb40-4a37-9133-232ea354cc72.png">

**Comment by jaklan** [comment]
Just two cents from my side - I have just encountered the same issue when trying to create a custom Airflow `dag` decorator being just the base one with some validation logic applied to kwargs. `functools.wraps(dag)` was not enough to solve the issue for Pyright-based type checking in VS Code, so based on the above discussion and [that SO answer](https://stackoverflow.com/a/71968448/6534668), I came up with the following wrapper:

```python

from functools import wraps
from typing import Any, Callable, TypeVar
from typing import cast

from airflow.decorators import dag as dag_decorator
from typing_extensions import ParamSpec

P = ParamSpec("P")
T = TypeVar("T")


def wrap(type: Callable[P, T]) -> Callable[[Callable[..., Any]], Callable[P, T]]:
    def wrapper(func: Callable[..., Any]) -> Callable[P, T]:
        return cast(Callable[P, T], wraps(type)(func))

    return wrapper


@wrap(dag_decorator)
def dag(*args, **kwargs):
    validated_kwargs = validate(kwargs)
    return dag_decorator(*args, **validated_kwargs)
```

Seems to work (pic below), but what I find interesting - if I remove `cast()` from the wrapper and just replace it with `return wraps(type)(func)` or even `return func` - I don't see any difference, it works as well. I guess Pyright simply relies on wrapper type hints and `cast()` is not even needed in such case. Having said that, I believe it's an implementation detail of a specific type checker, so returning `cast()` is just a safer option.

<img width="714" alt="image" src="https://github.com/user-attachments/assets/d929860e-0483-4a79-bd04-4382695b66f3" />

**Reply by jonathanunderwood** [reply]
I hit a similar problem/need recently when trying to wrap a Dagster decorator wrapper, and this approach worked for me there, too, even though the Dagster decorator function in question was also overloaded. Details are at https://github.com/microsoft/pyright/issues/10436. Many thanks for sharing this solution. I actually think functools would benefit from an addition along these lines!
